js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+var flexbuffers = {};
+
+flexbuffers.BitWidth = {
+  WIDTH8: 0,
+  WIDTH16: 1,
+  WIDTH32: 2,
+  WIDTH64: 3,
+};
+
+flexbuffers.BitWidth.toByteWidth = (bitWidth) => {
+  return 1 << bitWidth;
+};
+
+flexbuffers.BitWidth.width = (value) => {
+  if (Number.isInteger(value)) {
+    const v = value < 0 ? value * -1 : value;
+    if (v >> 7 === 0) return flexbuffers.BitWidth.WIDTH8;
```

Comment:
Comments from others in my team (not me, I have no idea):
"That's just incorrect, e.g. for v = 0x100000001 which is not an 8-bit integer but would take the return here."

Reply:
C++ version:
https://github.com/google/flatbuffers/blob/9abb2ec2cc92008e65037854f0d236c746c9b263/include/flatbuffers/flexbuffers.h#L183-L193

Reply:
not sure why I felt a macro was necessary.. apologies :)

Reply:
Well `Number` type is really broken in JavaScript. The C++ way you described, does not work either ;). 

Good news, there is `BigInt` type, which works in a sane way. Bad news it is currently not supported by all browsers. Most notable Safari, on desktop and on iOS. But the version of macOS and iOS which are introduced yesterday and will be rolled out this fall include Safari which supports BigInt. Generally we need BigInt to to build a buffer, reading will work fine until we need to access a value which is stored as an 8 byte int / uint. If some one have a great idea for fixing it. You have my attention.

Reply:
Bigint is also going to be slow, as it will be implemented as a dynamically allocated object in many cases.

I have no idea how to solve these kinds of problems in JS. From what I hear, to do bit-twiddling, you can basically rely on that working for 32-bit integers.. any bigger integers and you'd have to split it up manually. That's what FlatBuffers does for 64-bit numbers.

Reply:
I removed BigInt / BigUInt conversions when checking for int / unit width and when reading int / uint it is used only if supported by the platform.